### PR TITLE
return non-zero exit code if there are errors

### LINKED ISSUE #150: Set the exit code
Return an error code if there are tidy errors or Rustfmt failed to format something, OK otherwise.

I think this is easy, not 100% sure of that.

### PR DESCRIPTION
**Pull request by matklad**
closes #150

@nrc please review

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -239,9 +263,17 @@ pub struct FormatReport {
 }
 
 impl FormatReport {
+    fn new() -> FormatReport {
+        FormatReport { file_error_map: HashMap::new() }
+    }
+
     pub fn warning_count(&self) -> usize {
         self.file_error_map.iter().map(|(_, ref errors)| errors.len()).fold(0, |acc, x| acc + x)
```
Comment (matklad) [inline src/lib.rs]: I am keen to rename `warning_count -> error_count`.

**Comment by nrc** [comment]
The code looks good, however, I think it is important to distinguish between formatting errors, parsing errors, and errors in operation, and return different codes. I.e., the caller should be able to distinguish between an error like can't open a file and a formatting error like producing an over-long line.

And you should document this somewhere (in the README, I guess).

**Comment by matklad** [comment]
@nrc this now returns `1`, `2`, or `3` depending on what exactly has failed. I'm not sure what exact numbers should we use here. Maybe it is safer to leave `1` to misc, and use `102`, `103` for "can't parse", "can't format" situations?

**Comment by nrc** [comment]
looks good, thanks for the changes.

I think 1, 2, 3 is OK for now, we can change them later. We should reconfirm that they are right before 1.0.

**Comment by jonhoo** [comment]
This interacts pretty poorly with rust.vim, which will bail out by not saving **and** printing out the entire file contents in the terminal when rustfmt exist with a non-zero return code (see rust-lang/rust.vim#75).

**Comment by matklad** [comment]
`rustfmt` exits with non zero for a reason. It means that it was unable to format everything correctly (and syntax errors prevent formatting). Can `rust.vim` be changed so that it reports `rustfmt` failures more subtly?

**Comment by jonhoo** [comment]
@matklad: I agree, the correct thing for rustfmt to do is to indicate that it encountered problems. When the problems are terminal (i.e., nothing could be done), then the right thing to do is definitely to exit with an error code. When the problems are really warnings (e.g., one line couldn't be formatted as it was too long, but otherwise formatting completed), I'm not sure that is the right behavior. That said, this is definitely a bug in `rust.vim` (which is why it's been filed). The temporary workaround is to set `g:rustfmt_fail_silently = 1`.

**Pull request merged**